idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+}
+
+func (t *TagKeyValues) Len() int {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		total += len(kv) + 1
+	}
+	return total
+}
+
+func (t *TagKeyValues) MarshalBinaryTo(b []byte) error {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		if kv == "=" || kv == "" {
+			continue
+		}
+		copy(b[total:], kv)
+		total += len(kv)
+		copy(b[total:], ";")
+		total++
+	}
+	return nil
+}
+
+func (t *TagKeyValues) UnmarshalBinary(b []byte) error {
+	if len(b) < 3 {
+		return nil
+	}
+	tags := strings.Split(string(b[:len(b)-1]), ";")
+	tmp := make([]TagKeyValue, len(tags))
+	t.KeyValues = tmp
+	for i, tag := range tags {
+		if strings.HasPrefix(tag, "name") {
+			continue
+		}
+		if tag == "=" || tag == "" {
+			log.Error("idx: Empty tag, ignoring: ", tag)
+			invalidTag.Inc()
+			continue
+		}
+		if strings.Contains(tag, ";") {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		eqPos := strings.Index(tag, "=")
+		if eqPos < 0 {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+		if err != nil {
+			log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			(*t).KeyValues[i].Key = key
+		}
+
+		value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+		if err != nil {
+			log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			(*t).KeyValues[i].Value = value
+		}
+	}
+	return nil
+}
+
+// MetricDefinition stores information which identifies a single metric
+type MetricDefinition struct {
+	Id    schema.MKey
+	OrgId uint32
+	// using custom marshalling for MetricName
+	// if there is another way we should explore that
+	Name       MetricName `msg:"name,extension"`
+	Interval   int
+	Unit       string
+	mtype      mType
+	Tags       TagKeyValues `msg:"tagkeyvalues,extension"`
+	LastUpdate int64
+	Partition  int32
+}
+
+// NameWithTags returns a string version of the MetricDefinition's name with
+// all of its tags in the form of 'name;key1=value1;key2=value2;key3=value3'
+func (md *MetricDefinition) NameWithTags() string {
+	var bld strings.Builder
+
+	md.Name.string(&bld)
+	sort.Sort(md.Tags.KeyValues)
+	for _, tag := range md.Tags.KeyValues {
+		key, err := IdxIntern.GetStringFromPtr(tag.Key)
+		if err != nil {
+			panic(fmt.Sprintf("idx: Failed to retrieve interned tag key: %v", err))
+		}
+		if key == "name" {
+			continue
+		}
+		bld.WriteString(";")
+		tag.string(&bld)
+	}
+	return bld.String()
+}
+
+// NameWithTagsHash returns an Md5Hash struct containing the
+// hashed md5 sum of a NameWithTags for the given MetricDefinition
+func (md *MetricDefinition) NameWithTagsHash() Md5Hash {
+	md5Sum := md5.Sum(bytes.NewBufferString(md.NameWithTags()).Bytes())
```

Comment:
wouldn't `[]bytes(md.NameWithTags())` be the same as `bytes.NewBufferString(md.NameWithTags()).Bytes()`

Reply:
yes, `[]byte(md.NameWithTags())` would be the same in this case.